### PR TITLE
[Cache] Add PhpArrayAdapter to use shared memory on PHP 7.0

### PR DESCRIPTION
**Pull request by tgalopin**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | WIP |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Depends on https://github.com/symfony/symfony/pull/18825

This PR aims to implement a Symfony Cache adapter able to build files optimized for OPCache memory storage. Any kind of static data that can be warmed up is a candidate for OPcache storage.

This PR is limited to the implementation of the Adapter. It's usage in the annotations caching system is implemented in the PR https://github.com/symfony/symfony/pull/18533. The aim is also to use this adapter in other contexts (validator, serializer, ...).

### CODE REVIEW

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+        if (!isset($this->values[$key])) {
+            return $this->fallbackPool->getItem($key);
+        }
+
+        $value = $this->values[$key];
+
+        if ('N;' === $value) {
+            $value = null;
+        } elseif (isset($value[2]) && is_string($value) && ':' === $value[1]) {
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: ``is_string`` should come before ``isset($value[2])`` IMO

File: src/Symfony/Component/Cache/Adapter/NullAdapter.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    private function generateItems(array $keys)
+    {
+        $f = $this->createCacheItem;
+
+        foreach ($keys as $key) {
+            yield $key => $f($key);
+        }
+    }
+}
```
Comment (phansys) [inline src/Symfony/Component/Cache/Adapter/NullAdapter.php]: Missing LF at EOF.

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+            if (!is_dir($directory) && !@mkdir($directory, 0777 & (~$this->umask), true)) {
+                throw new InvalidArgumentException(sprintf('Cache directory does not exist and cannot be created: %s', $directory));
+            }
+
+            if (!is_writable($directory)) {
+                throw new InvalidArgumentException(sprintf('Cache directory is not writable: %s', $directory));
+            }
+        }
+
+        $dump = <<<EOF
```
Comment (phansys) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: This should be enclosed with nowdoc instead of heredoc.

File: src/Symfony/Component/Cache/Adapter/NullAdapter.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class NullAdapter implements AdapterInterface
+{
+    private $createCacheItem;
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Adapter/NullAdapter.php]: phpdoc missing

File: src/Symfony/Component/Cache/Adapter/NullAdapter.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class NullAdapter implements AdapterInterface
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Adapter/NullAdapter.php]: i think this `NullAdapter` belongs to another PR which you already did, or am i wrong?

#18825

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+class OpCacheAdapter implements AdapterInterface, LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private $file;
+    private $umask;
+    private $values;
+    private $createCacheItem;
+
+    /** @var AdapterInterface */
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: don't know the symfony standard, but i would write

```php
/**
 * @var AdapterInterface
 */
```

instead of one line

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+ * You can use the method "store" to build the cache file.
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class OpCacheAdapter implements AdapterInterface, LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private $file;
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: do we need php doc for private properties

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+            }
+        }
+
+        if ($fallbackKeys) {
+            foreach ($this->fallbackPool->getItems($fallbackKeys) as $key => $item) {
+                yield $key => $item;
+            }
+        }
+    }
+}
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: Missing line ending

File: src/Symfony/Component/Cache/Tests/Adapter/NullAdapterTest.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Tests\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\Adapter\NullAdapter;
+
+class NullAdapterTest extends \PHPUnit_Framework_TestCase
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Tests/Adapter/NullAdapterTest.php]: this belongs to the other PR as well

File: src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterTest.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        call_user_func(\Closure::bind(function () use ($item) {
+            $this->values[$item->getKey()] = $item->get();
+            $this->warmUp($this->values);
+            $this->values = eval(substr(file_get_contents($this->file), 6));
+        }, $this, OpCacheAdapter::class));
+
+        return true;
+    }
+}
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterTest.php]: Missing line ending

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php (hunk 7 above)
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: not when the constructor allows IDE to infer their type and when the name is clear enough (and if the name is not clear, the preferred way is to use a better name)

File: src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterWithFallbackTest.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+    public function createCachePool()
+    {
+        if (defined('HHVM_VERSION')) {
+            $this->skippedTests['testDeferredSaveWithoutCommit'] = 'Fails on HHVM';
+        }
+
+        return new OpCacheAdapter(self::$file, new FilesystemAdapter('opcache-fallback'));
+    }
+}
```
Comment (OskarStark) [inline src/Symfony/Component/Cache/Tests/Adapter/OpCacheAdapterWithFallbackTest.php]: missing line ending

File: src/Symfony/Component/Cache/Adapter/NullAdapter.php (hunk 5 above)
Comment (tgalopin) [inline src/Symfony/Component/Cache/Adapter/NullAdapter.php]: This PR is based on #18825 which means that when #18825 will be merged I'll rebase this one on master to remove the commit. In the meantime, I use the commit for the tests.

File: src/Symfony/Component/Cache/Tests/Adapter/NullAdapterTest.php (hunk 9 above)
Comment (tgalopin) [inline src/Symfony/Component/Cache/Tests/Adapter/NullAdapterTest.php]: Same reason

**Comment by tgalopin** [comment]
I did the changes

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+
+// This file has been auto-generated by the Symfony Cache Component.
+
+return array(
+
+
+EOF;
+
+        foreach ($values as $key => $value) {
+            CacheItem::validateKey($key);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: because PHP casts numeric string keys to ints, we should cast ints to string here:
`is_int($key) ? (string) $key : $key);

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
+
+        if (!isset($this->values[$key])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: let's validate before that $key is a string to prevent issues with magic array index casting (same below)

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+     *
+     * @return \Generator
+     */
+    private function generateItems(array $keys)
+    {
+        $f = $this->createCacheItem;
+        $fallbackKeys = array();
+
+        foreach ($keys as $key) {
+            if (!is_string($key)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: this check should be done before calling generateItems

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: missing is_string check

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deleteItem($key)
+    {
+        if (null === $this->values) {
+            $this->initialize();
+        }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: missing is_string check

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+
+        $deleted = true;
+        $fallbackKeys = array();
+
+        foreach ($keys as $key) {
+            if (isset($this->values[$key])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: missing is_string check

File: src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php (hunk 14 above)
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/OpCacheAdapter.php]: indeed. Doing it in the generator will throw the exception too late (the iterator will throw it during iteration, while it must be throw by getItems)

**Comment by nicolas-grekas** [comment]
I suggest renaming this adapter to Php7Adapter (ans use the OpcacheAdapter name in #18894)

**Comment by tgalopin** [comment]
I did the changes

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by nicolas-grekas** [comment]
Btw I looked at this umask issue and it happens that Symfony considers this as an ops issue that Symfony should not deal with. Let's remove it from this PR.

File: src/Symfony/Component/Cache/Adapter/Php7Adapter.php [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+        $f = $this->createCacheItem;
+        $fallbackKeys = array();
+
+        foreach ($keys as $key) {
+            if (isset($this->values[$key])) {
+                $value = $this->values[$key];
+
+                if ('N;' === $value) {
+                    $value = null;
+                } elseif (isset($value[2]) && is_string($value) && ':' === $value[1]) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Php7Adapter.php]: is_string should be first

File: src/Symfony/Component/Cache/Adapter/Php7Adapter.php [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+            return new static($file, $fallbackPool);
+        }
+
+        return $fallbackPool;
+    }
+
+    /**
+     * @param string           $file         The PHP file were values are cached.
+     * @param AdapterInterface $fallbackPool A pool to fallback on when an item is not hit.
+     * @param int              $umask
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Php7Adapter.php]: IMHO we should drop this umask concern: there is no other place in the Symfony code base where we deal with this, which means that we've always considered that a wrong system umask (zero), is an ops mistake, and it's not Symfony's job to try to fix external mistakes. We've had the same reasonings about e.g. Varnish misconfigs.

**Comment by tgalopin** [comment]
I did the changes

**Comment by nicolas-grekas** [comment]
I renamed #18894 back to PhpFilesAdapter, and I now think this one should be renamed PhpArrayAdapter. Easier to understand, more explicit complementarity.

**Comment by tgalopin** [comment]
I updated this

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * Adapter building static PHP files that will be cached by OPCache.
+ * This Adapter is in read-only if you use AdapterInterface methods.
+ * You can use the method "store" to build the cache file.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: store -> warmUp

**Comment by nicolas-grekas** [comment]
Calling for votes @symfony/deciders

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+     */
+    private $createCacheItem;
+
+    /**
+     * @var AdapterInterface
+     */
+    private $fallbackPool;
+
+    /**
+     * @param string $file The PHP file were values are cached.
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: We need a description of this method usage and it should be moved after the constructor.

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+     *
+     * @return CacheItemPoolInterface
+     */
+    public static function createOnPhp7($file, CacheItemPoolInterface $fallbackPool)
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: I don't like the method name. I suppose that this will also work on PHP 8, right? So including a version in a method name sounds fragile to me. But then, I don't understand the usage, so I might be wrong.

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: I suggest using simply `create` and ensure the doc block is clear enough about it's behavior

**Comment by tgalopin** [comment]
I updated this

**Comment by tgalopin** [comment]
I fixed a problem with HHVM

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+        $this->createCacheItem = \Closure::bind(
+            function ($key, $value) {
+                $item = new CacheItem();
+                $item->key = $key;
+                $item->value = $value;
+                $item->isHit = true;
+
+                return $item;
+            },
+            $this,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: `null` instead of `$this`

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * Adapter building static PHP files that will be cached by OPCache.
+ * This Adapter is in read-only if you use AdapterInterface methods.
+ * You can use the method "warmUp" to build the cache file.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: I suggest:
```
 * Caches items at warm up time using a static PHP array that is stored in shared memory by OPCache since PHP 5.6.
 * Warmed up items are read-only and run-time discovered items are cached using a fallback adapter.
```

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php (hunk 23 above)
Comment (tgalopin) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Well, shouldn't we remove the bind then?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: the next line is the important one

**Comment by fabpot** [comment]
@tgalopin Can you take comments into account? Looks like we are almost there :)

**Comment by tgalopin** [comment]
I did some benchmarks to check the PHP 5.6 support in this PR.

I used three Docker containers with only PHP/HHVM installed, on the same host. I used Apache Benchmark with 1000 requests at a concurrency of 100 (`ab -n 1000 -c 100`).

The results are the following:

| Engine    	| Context              	  	| Time per request 	|
| -----------		| ------------------------		| ------------------		|
| PHP 5.6   	| Empty array            	| 0.953 ms/req     	|
| PHP 5.6   	| Array with 15 000 keys 	| 4.671 ms/req     	|
| PHP 7.0   	| Empty array            	| 0.688 ms/req     	|
| PHP 7.0   	| Array with 15 000 keys 	| 0.709 ms/req     	|
| HHVM 3.12 	| Empty array            	| 0.869 ms/req     	|
| HHVM 3.12 	| Array with 15 000 keys 	| 0.875 ms/req     	|

This seems to indicate that having an array filled with data in PHP 5.6 is significantly slower than having an empty array. Therefore, I removed the PHP 5.6 support for this PhpArrayAdapter in the `create` factory method (and I added the HHVM support).

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by dunglas** [comment]
👍

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+EOF;
+
+        foreach ($values as $key => $value) {
+            CacheItem::validateKey(is_int($key) ? (string) $key : $key);
+
+            if (null === $value || is_object($value)) {
+                try {
+                    $value = serialize($value);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable %s value', $key, get_class($value)), 0, $e);
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Missing closing double quote.

Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Also missing a dot at the end.

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 26]
```diff
@@ -0,0 +1,358 @@
...
+                }
+            } elseif (is_array($value)) {
+                try {
+                    $serialized = serialize($value);
+                    $unserialized = unserialize($serialized);
+                } catch (\Exception $e) {
+                    throw new InvalidArgumentException(sprintf('Cache key "%s has non-serializable array value', $key), 0, $e);
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: missing dot at the end (and probably on other exception messages).

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 27]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        $this->values = array();
+
+        @unlink($this->file);
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Don't we have to return `false` too if it was not possible to delete the file?

Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: Let's do by using `file_exists` after unlinking when `unlink` returns `false`. Even if it opens a race condition, warming up the cache should be rare and slower than the time between the `unlink` and `file_exists`, so that in reality the race doesn't happen.

**Comment by tgalopin** [comment]
I did the changes.

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php [hunk 28]
```diff
@@ -0,0 +1,358 @@
...
+ */
+class PhpArrayAdapter implements AdapterInterface
+{
+    private $file;
+    private $values;
+    private $createCacheItem;
+    private $fallbackPool;
+
+    /**
+     * @param string           $file         The PHP file were values are cached.
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: dots at the end of @param (@return, @throws, ... as well) should be removed

**Comment by tgalopin** [comment]
I did the change

**Comment by fabpot** [comment]
Thank you @tgalopin.
